### ISSUE TITLE
Velocity annotation

### ISSUE THREAD

**Issue by johnchamberlin** [sdparekh/zUMIs#135]
Can you clarify how zUMIs identifies "spliced" and "unspliced" UMIs for the velocity feature? The exon counts and the intron counts are not equivalent to these categories because the exon counts include intron-exon junction reads. But the difference might be a trivial. Similarly, some UMIs have both intron and exon reads.

thanks,

**Comment by cziegenhain** [comment]
Hi,

For the velocity is not calculated by functions we wrote, but just is a convenience function around the standard [velocyto package](http://velocyto.org/velocyto.py/index.html).

For counting in zUMIs: 
Reads that are overlapping junctions and are partially exon and intron, zUMIs will always consider these as exonic, as [described in our paper](https://academic.oup.com/gigascience/article/7/6/giy059/5005022#117719929).

Hope this helps!

**Comment by johnchamberlin** [comment]
So the zUMIs-generated BAM file is fed into Velocyto?

**Comment by cziegenhain** [comment]
Exactly. We just rewrite the bam file to swap the cell barcode tag (from BC to CB) to make it compatible, but that's a simple step.

**Comment by johnchamberlin** [comment]
thanks!

**Comment by LiBuchauer** [comment]
> Exactly. We just rewrite the bam file to swap the cell barcode tag (from BC to CB) to make it compatible, but that's a simple step.

Hi Christoph, 

if I only have access to zUMI output files generated by the pre-velocyto-compatible version of zUMIs (i.e. '.aligned.sorted.bam', '.aligned.sorted.bam.ex.featureCounts', '.aligned.sorted.bam.in.featureCounts'), e.g. because I am a downstream collaborator ;-), can I still make one of them ('.aligned.sorted.bam'?) compatible for velocyto spliced/unspliced mapping by renaming the tag myself?

Thanks a lot

Lisa

**Comment by LiBuchauer** [comment]
It now seems to me like there are no barcodes in this bam file at all (?sorry, beginner with BAM/SAM). Would it be an option to retrieve them from the corresponding '.barcodelist.filtered.sort.sam' and somehow fudge it together?

**Comment by cziegenhain** [comment]
Hi Lisa,

Sorry with this old zUMIs output it will be hard to piece together. Please ask your collaborators to share the fastq files so you can use a new version of the pipeline.

Best,
Christoph

**Comment by LiBuchauer** [comment]
Hi Christoph,

in the meantime I (maybe) figured it out, velocyto still doesn't run because of samtools issues, so I am not 100% sure the file is okay, but everything crashes later now than it did previously and there are no complaints because of the BAM file ;-).

Anyways, it may be the most inefficient piece of code for sam manipulation ever, but for reference of others interested in the topic:

```python
import pysam
def make_bam_with_barcodes(
        bampath,
        sampath,
        outfolder):
    """ Based on the zUMI ouputs  '*.aligned.sorted.bam', which does not
    contain cell and molecular barcodes, and '*.barcodelist.filtered.sort.sam',
    which contains only barcodes, creates a new bam file that contains
    cell barcode as tag 'CB' and molecular barcode as tag 'UB'. Preserves the
    order of the original files (which is identical for both files).
    Output is stored in the given folder with a filename resembling the input
    bam file's name."""
    # TODO: implement check that both files stem from the same experiment,
    # ideally also that they have the same length

    # import sam and bam file
    samo = pysam.AlignmentFile(sampath, 'r', check_sq=False)
    bamo = pysam.AlignmentFile(bampath, 'rb')

    # construct name for the output
    # name of input bam file
    name_in = bampath.split('/')[-1].split('.')[0]
    outpath = outfolder + '/' + name_in + '.aligned.sorted.BARCODES.bam'

    # now, open a new bam file and copy the header from the old one
    outfile = pysam.AlignmentFile(outpath, 'wb', header=bamo.header)

    # go through all reads, add barcodes and write them to the new file
    for s, b in zip(samo.fetch(until_eof=True), bamo.fetch(until_eof=True)):
        # make a new read object
        a = pysam.AlignedSegment()

        # copy all the information from the original bam read
        a = b
        # but: we want to alter the tags and add new info from the same file!
        # get combined cell and molecular barcode
        combi = s.query_sequence
        CB = combi[:7]  # cell barcode is stored in first 7 characters here
        UB = combi[7:]  # molecular barcode is stored in last 8 characters
        # now get the tag list (enforce list, could be set) and add the new
        # BC and UB tags
        taglist = list(b.tags)
        taglist += [('CB', CB), ('UB', UB)]
        # write this to the new read object
        a.tags = taglist

        # finally, write this read to the new bamfile
        outfile.write(a)

    # close all files
    samo.close()
    bamo.close()
    outfile.close()```

Will update if I get it to run.

**Comment by LiBuchauer** [comment]
Update: For me it worked. In brief, I used old zUMI V1.4 output because this is what I had, I made a new bamfile that complied with the velocyto requirements using barcodes from the correspnding samfile, I then had to sort (samtools sort) this, and then velocyto ran without issues, producing a loomfile with (after rough analysis) meaningful content.

Thanks for all the comments and support you provided over all the issues in this repository, that really helped a lot.
